### PR TITLE
Support directory-based dependencies

### PR DESCRIPTION
**Pull request by abompard**
## Related Issues and Dependencies

Fixes: #203

## This introduces a breaking change

- [ ] Yes
- [x] No

## This Pull Request implements

This pull request adds support for poetry projects that have local dependencies. Previously, only the `git` source type was supported.

### CODE REVIEW

**Comment by fridex** [comment]
@abompard some tests failed as a new pip has been released. Could you please rebase on top of https://github.com/thoth-station/micropipenv/pull/205? Thanks again for the patch 👍🏻

**Comment by frenzymadness** [comment]
Sorry it took so long I was on PTO last week. I like the code, thanks for the new feature. The tests are also good but I'd add also some installability tests. The current tests test that micropipenv is able to generate a pipfile.lock from poetry.lock and pyptoject.toml. It'd be also nice to test that it's able to install a project from a directory. What do you think?

**Comment by abompard** [comment]
> It'd be also nice to test that it's able to install a project from a directory. What do you think?

Sounds good, I'll try to find a way to test that.

**Comment by frenzymadness** [comment]
Is this ready to be reviewed?

**Comment by abompard** [comment]
I think so, I've added the tests that you asked for.

**Comment by frenzymadness** [comment]
Could you please rebase it on top of the master branch? #207 fixed an issue with the latest release of flexmock.

**Comment by abompard** [comment]
Done!

**Comment by abompard** [comment]
I'm not sure why the Windows test is failing, it seems unrelated to my changes. Could someone check it out please?

**Comment by frenzymadness** [comment]
I see that one test on Linux fails - and because all of them passed on master branch 4 days ago, it's either an unstable test or it has something to do with the changes here.

The test tries to install from this file:
```
-e .[foo]
daiquiri==2.0.0 \
    --hash=sha256:6b235ed15b73b87fd3cc2521aacbb727bf8443a0896dc534b07503841d03cfdb \
    --hash=sha256:d57b9fd5432933c6e899054eb62cee22eab89f560c8493254d327ec27893c866
python-json-logger==0.1.11 \
    --hash=sha256:b7a31162f2a01965a5efb94453ce69230ed208468b0bbc7fdfc56e6d8df2e281 \
```
which is then translated to:
```
2021-11-10T11:46:41.3570552Z ---------------------------------- Pipfile.lock ----------------------------------
2021-11-10T11:46:41.3571377Z {
2021-11-10T11:46:41.3571732Z     "_meta": {
2021-11-10T11:46:41.3572098Z         "hash": {
2021-11-10T11:46:41.3573407Z             "sha256": "cab3f5e8ccb3907fc1e48407550639c3fe9a43551e74b32e2e3cf94b6a2de0e1"
2021-11-10T11:46:41.3574408Z         },
2021-11-10T11:46:41.3575032Z         "pipfile-spec": 6,
2021-11-10T11:46:41.3575638Z         "requires": {
2021-11-10T11:46:41.3576088Z             "python_version": "3.10"
2021-11-10T11:46:41.3576501Z         },
2021-11-10T11:46:41.3576856Z         "sources": [
2021-11-10T11:46:41.3577232Z             {
2021-11-10T11:46:41.3578061Z                 "name": "5a06749b2297be54ac5699f6f2761716adc5001a2d5f8b915ab2172922dd5706",
2021-11-10T11:46:41.3579082Z                 "url": "https://pypi.org/simple",
2021-11-10T11:46:41.3579656Z                 "verify_ssl": true
2021-11-10T11:46:41.3580048Z             }
2021-11-10T11:46:41.3580395Z         ]
2021-11-10T11:46:41.3580713Z     },
2021-11-10T11:46:41.3581081Z     "default": {
2021-11-10T11:46:41.3582061Z         "9fc5eb29b17d3a33a8f36a5d6351e8567b42f4d37e404ec6c887e7abb9cf3b17": {
2021-11-10T11:46:41.3583085Z             "editable": true,
2021-11-10T11:46:41.3583826Z             "path": "file:///github/workspace/tests/data/install/requirements_editable"
2021-11-10T11:46:41.3584463Z         },
2021-11-10T11:46:41.3584849Z         "daiquiri": {
2021-11-10T11:46:41.3585248Z             "hashes": [
2021-11-10T11:46:41.3586256Z                 "sha256:6b235ed15b73b87fd3cc2521aacbb727bf8443a0896dc534b07503841d03cfdb",
2021-11-10T11:46:41.3587781Z                 "sha256:d57b9fd5432933c6e899054eb62cee22eab89f560c8493254d327ec27893c866"
2021-11-10T11:46:41.3588697Z             ],
2021-11-10T11:46:41.3589549Z             "index": "5a06749b2297be54ac5699f6f2761716adc5001a2d5f8b915ab2172922dd5706",
2021-11-10T11:46:41.3590410Z             "version": "==2.0.0"
2021-11-10T11:46:41.3590798Z         },
2021-11-10T11:46:41.3591434Z         "python-json-logger": {
2021-11-10T11:46:41.3591944Z             "hashes": [
2021-11-10T11:46:41.3592907Z                 "sha256:b7a31162f2a01965a5efb94453ce69230ed208468b0bbc7fdfc56e6d8df2e281"
2021-11-10T11:46:41.3594069Z             ],
2021-11-10T11:46:41.3594919Z             "index": "5a06749b2297be54ac5699f6f2761716adc5001a2d5f8b915ab2172922dd5706",
2021-11-10T11:46:41.3595791Z             "version": "==0.1.11"
2021-11-10T11:46:41.3596190Z         }
2021-11-10T11:46:41.3596509Z     },
2021-11-10T11:46:41.3596883Z     "develop": {}
2021-11-10T11:46:41.3597237Z }
```
And then, micropipenv and pip produces this:
```
2021-11-10T11:46:43.1644439Z 11-10-21 11:46:43 micropipenv  INFO   Installing '9fc5eb29b17d3a33a8f36a5d6351e8567b42f4d37e404ec6c887e7abb9cf3b17'
2021-11-10T11:46:43.4436685Z ERROR: /github/workspace/tests/data/install/requirements_editable/file:/github/workspace/tests/data/install/requirements_editable is not a valid editable requirement. It should either be a path to a local project or a VCS URL (beginning with svn+, git+, hg+, or bzr+).
2021-11-10T11:46:43.4799601Z Traceback (most recent call last):
2021-11-10T11:46:43.4800492Z   File "/github/workspace/micropipenv.py", line 1280, in <module>
2021-11-10T11:46:43.4805104Z     return_code = main()
2021-11-10T11:46:43.4805910Z   File "/github/workspace/micropipenv.py", line 1266, in main
2021-11-10T11:46:43.4811320Z     handler(**arguments)
2021-11-10T11:46:43.4811988Z   File "/github/workspace/micropipenv.py", line 884, in install
2021-11-10T11:46:43.4817344Z     install_requirements(pip_bin, pip_args=pip_args)
2021-11-10T11:46:43.4818433Z   File "/github/workspace/micropipenv.py", line 819, in install_requirements
2021-11-10T11:46:43.4823353Z     install_pipenv(pip_bin, pipfile_lock=pipfile_lock, pip_args=pip_args, deploy=False)
2021-11-10T11:46:43.4824369Z   File "/github/workspace/micropipenv.py", line 364, in install_pipenv
2021-11-10T11:46:43.4827505Z     raise PipInstallError(
2021-11-10T11:46:43.4831211Z __main__.PipInstallError: Failed to install requirements, dependency '9fc5eb29b17d3a33a8f36a5d6351e8567b42f4d37e404ec6c887e7abb9cf3b17' could not be installed
2021-11-10T11:46:43.6448436Z FAILED
```
It seems to be related to the new code from this PR.

**Comment by abompard** [comment]
Oh good catch, I didn't see the linux failure. There was a bug indeed, I fixed it, hopefully the tests will pass now.

**Comment by abompard** [comment]
Ah this time I don't think it's me, apparently pip in git requires Python `>=3.7`

**Comment by frenzymadness** [comment]
> Ah this time I don't think it's me, apparently pip in git requires Python `>=3.7`:
> 
> ```
> ERROR: Package 'pip' requires a different Python: 3.6.15 not in '>=3.7'
> ```

Oh my. I'm gonna fix this and then we can rebase this PR again.

**Comment by abompard** [comment]
Yeah, that's the cause : https://github.com/pypa/pip/commit/0252c04a16cd93fe422cebf0b48453b559a2e404

**Comment by frenzymadness** [comment]
I'm sorry for the delay, PR #209 should fix the CI and then we can get back to this.

**Comment by frenzymadness** [comment]
CI should be fixed in master. Could you please rebase this PR?

**Comment by abompard** [comment]
@frenzymadness Done!

**Comment by abompard** [comment]
Could someone approve the workflows runs please?

**Comment by frenzymadness** [comment]
Tests are running. I'm really sorry that it takes so long but I'm currently on PTO. I'll try to get to this as soon as possible.

**Comment by frenzymadness** [comment]
There seems to be a problem on Windows where the path to the editable dependency is not escaped properly and therefore looks like this:
```
2021-11-25T15:41:42.6911452Z D:amicropipenvmicropipenvtestsdatainstallpipenv_editable should either be a path to a local project or a VCS url beginning with svn+, git+, hg+, or bzr+
```
I don't have access to a windows machine right now to debug it. Do you?

**Comment by abompard** [comment]
I don't either, I'll try to find what causes it, because it's not in one of my tests

**Comment by abompard** [comment]
Oh I think I found why, I'm parsing what can be a filesystem path with `urlparse()`. Fixing.

**Comment by abompard** [comment]
Ah, it failed again but I think I fixed it right this time, sorry it takes a couple iterations without a Windows box to check. Pip actually [uses forwards slashes](https://packaging.python.org/tutorials/installing-packages/#installing-from-local-archives) even on Windows. The new method I used should be less intrusive (no `abspath`).

**Comment by frenzymadness** [comment]
It seems it's time to do a review! Good job!

**Review by frenzymadness — COMMENTED** [review]
I like it, thanks a lot for your effort. I have just two minor comments.

File: micropipenv.py
Comment (abompard) [reply micropipenv.py:985]: Yeah I agree, it's not great. I've added a comment and small testcase for the formats I expect to see in this `path` option. What do you think?

**Comment by frenzymadness** [comment]
Looks good to me! Thanks a lot. Not approving yet to prevent automatic merge because I want to ask: @fridex it's a kinda big change - do you want to do a second review?

File: micropipenv.py
Comment (fridex) [inline micropipenv.py:987]: Not sure if I read this correctly - would this work even if the path is a directory starting with a dot - e.g. `.mylib`? Probably not that common, but just to clarify.

**Review by fridex — COMMENTED** [review]
Changes look good to me, I just have one (maybe too nitpicking) question. Thanks a lot for your work 👍🏻 💯

File: micropipenv.py
Comment (frenzymadness) [reply micropipenv.py:987]: That, from my experience, depends on pip. Whatever is a valid package name, pip tries to find it in the index. But if it contains "/", it's no longer a valid name so it tries to find it locally. It seems that dot is not allowed either so ".testproject" should work.

Comment (fridex) [reply micropipenv.py:987]: Ah, the result is passed to pip as an argument. I missed the bigger picture here. Thanks for the clarification. 👍🏻

**Review by fridex — APPROVED** [review]
lgtm, thanks for your time and contribution 👍🏻

**Review by frenzymadness — APPROVED** [review]

**Comment by fridex** [comment]
It looks like prow is down. Merging manually. CC @harshad16

**Pull request merged**
